### PR TITLE
 initial tgw support

### PR DESCRIPTION
**Pull request by adisos**
issue #164

### CODE REVIEW

File: pkg/ibmvpc/parser.go [hunk 1]
```diff
@@ -741,6 +795,133 @@ func getNACLconfig(rc *ResourcesContainer,
...
+					fmt.Printf("warning: ignoring TGW %s, as currently not supporting connected VPCs with overlapping address ranges\n",
+						tgw.ResourceName)
+					continue
+				}
+				vpcsAddressRanges = vpcsAddressRanges.Union(vpcConfig.VPC.(*VPC).internalAddressRange)
+			}
+
+			// simple union of vpc zones
+			for name, zone := range vpc.zones {
+				// currently assuming no same zones names in the different vpcs
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:888]: How can this affect the results?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/vpcmodel/abstractVPC.go [hunk 2]
```diff
@@ -57,14 +57,16 @@ type Node interface {
 	Cidr() string
 	IsInternal() bool
 	IsPublicInternet() bool
+	VPC() VPCResourceIntf // the VPC to which this node belongs to
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/abstractVPC.go]: yes, makes sense -- changed

File: pkg/ibmvpc/parser.go
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:1]: opened issue #256

File: pkg/ibmvpc/parser.go [hunk 3]
```diff
@@ -741,6 +795,133 @@ func getNACLconfig(rc *ResourcesContainer,
 	return nil
 }
 
+func getTgwConfig(c *ResourcesContainer,
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: changed

File: pkg/ibmvpc/parser.go [hunk 4]
```diff
@@ -741,6 +795,133 @@ func getNACLconfig(rc *ResourcesContainer,
...
+			},
+			internalAddressRange: vpcsAddressRanges,
+			nodes:                internalNodes,
+			zones:                vpcsZones,
+		}
+		nacls.vpc = newConfig.VPC.(*VPC)
+		sgs.vpc = newConfig.VPC.(*VPC)
+		// TODO: analysis should warn if more than one naclLayer/sgLayer is present in FilterTrafficResource, as it is going
+		// to be ignored
+		newConfig.FilterResources = append(newConfig.FilterResources, []vpcmodel.FilterTrafficResource{nacls, sgs}...)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go [hunk 5]
```diff
@@ -741,6 +795,133 @@ func getNACLconfig(rc *ResourcesContainer,
...
+					return fmt.Errorf("unexpected type for filter resource in VPC %s", vpc.ResourceUID)
+				}
+			}
+
+			// omit routing resources -- assuming only internal vpc-to-vpc connectivity is of interest to analyze
+			// TODO: is there a scenario of connectivity from one vpc's vsi to external entity through another vpc's pgw/fip ?
+
+			// simple union for NameToResource map
+			for n, r := range vpcConfig.NameToResource {
+				newConfig.NameToResource[n] = r
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:908]: opened issue #256

File: pkg/ibmvpc/parser.go (hunk 1 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: It cannot, removed (left from older version of the code..)

File: pkg/ibmvpc/parser.go [hunk 6]
```diff
@@ -741,6 +795,133 @@ func getNACLconfig(rc *ResourcesContainer,
...
+				ResourceUID:  vpcPrefix + tgw.ResourceUID,
+				ResourceType: ResourceTypeVPC,
+			},
+			internalAddressRange: vpcsAddressRanges,
+			nodes:                internalNodes,
+			zones:                vpcsZones,
+		}
+		nacls.vpc = newConfig.VPC.(*VPC)
+		sgs.vpc = newConfig.VPC.(*VPC)
+		// TODO: analysis should warn if more than one naclLayer/sgLayer is present in FilterTrafficResource, as it is going
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go:945]: The assumption is that there is at most one naclLayer and one sgLayer objects in FilterTrafficResource, but it is currently not checked.

File: pkg/vpcmodel/textOutput.go [hunk 7]
```diff
@@ -1,20 +1,30 @@
...
 	case AllEndpoints, AllSubnets, SingleSubnet:
-		return fmt.Sprintf("Analysis for VPC %s\n", vpcName)
+		if c1.IsMultipleVPCsConfig {
+			if len(c1.RoutingResources) != 1 {
+				return "", errors.New("unexpected config of multiple VPCs connected by TGW, missing TGW resource")
+			}
+			tgw := c1.RoutingResources[0]
+			return fmt.Sprintf("Analysis for connectivity between VPCs connected by TGW %s (UID: %s)\n", tgw.Name(), tgw.UID()), nil
+		}
+		return fmt.Sprintf("Analysis for VPC %s\n", vpcName), nil
```
Comment (zivnevo) [inline pkg/vpcmodel/textOutput.go:23]:
```suggestion
		return fmt.Sprintf("Connectivity for VPC %s\n", vpcName), nil
```

**Review by zivnevo — CHANGES_REQUESTED** [review]

**Review by zivnevo — APPROVED** [review]

**Review by zivnevo — APPROVED** [review]

**Pull request merged**